omod/src/main/java/org/openmrs/module/web/filter/AdminAuthorisationFilter.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.web.filter;
+
+import java.io.IOException;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.openmrs.User;
+import org.openmrs.api.context.Context;
+import org.openmrs.web.WebConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This filter checks if an authenticated user trying to access administrative functions has the
+ * <STRONG>System Administration<STRONG> privelege. It will intercept any requests with *admin/* in
+ * its url. Unauthorised user will be redirected to the home page.
+ */
+public class AdminAuthorisationFilter implements Filter {
+
+	private static final Logger log = LoggerFactory.getLogger(AdminAuthorisationFilter.class);
+
+	/**
+	 * @see Filter#init(FilterConfig)
+	 */
+	public void init(FilterConfig filterConfig) throws ServletException {
+
+	}
+
+	/**
+	 * @see javax.servlet.Filter#doFilter(javax.servlet.ServletRequest,
+	 *      javax.servlet.ServletResponse, javax.servlet.FilterChain)
+	 */
+	public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain) throws IOException, ServletException {
+		HttpServletRequest httpReq = (HttpServletRequest) req;
+		User user = Context.getAuthenticatedUser();
+
+		if (!(user.isSuperUser() || user.hasRole("Organizational: System Administrator") || user.hasRole("Organizational: Hospital Administrator"))) {
+			httpReq.getSession().setAttribute(WebConstants.DENIED_PAGE, httpReq.getRequestURI());
```

Comment:
I think we have to revert to checking for only `App: coreapps.systemAdministration`, like you had before.

To resolve the issue that @dkayiwa mentioned, let's just check for the existence of the `App: coreapps.systemAdministration` privilege before we do anything else. So there will be three cases:

1. The `App: coreapps.systemAdministration` privilege exists, and the user has it. Therefore, the user should be allowed to view the /admin/* pages
2. The `App: coreapps.systemAdministration` privilege exists, and the user does not have it. Therefore, the user should be prevented from viewing the /admin/* pages
3. The `App: coreapps.systemAdministration` does not exist, because the implementation is legacy-ui only. Therefore, we should let the user view the /admin/* pages to maintain backward compatibility.

Reply:
This looks very clear to me @isears and am sure it will work for us.  I need to come up with a way of checking if the system has the `App: coreapps.systemAdministration ` privilege